### PR TITLE
Fix parsing of time_strings lacking leading zeroes

### PR DESCRIPTION
**Pull request by meshula**
**Link the Issue(s) this Pull Request is related to.**

#1293

@jchen9 Hi Julian, here's a rework of the existing routine to address the problem you found, and introduce a little more rigor to what is or is not a well-formed time string.

```Fixes #1293```

**Summarize your change.**

Describe the reason for the change.

- fixes parsing of time strings without leading zeroes.
- enforces that a negative sign can only appear in the left most position
- implementation does not allocate memory or copy strings
- implementation does not allow exponential notation and other things that std does allow but are inappropriate for time strings
-  compatible with strings produced by ffprobe

**Reference associated tests.**

adds C based tests corresponding to the existing Python based tests.

### CODE REVIEW

File: tests/test_opentime.cpp [hunk 1]
```diff
@@ -38,6 +38,75 @@ main(int argc, char** argv)
         assertFalse(t1 != t3);
     });
 
+    tests.add_test("test_from_time_string", [] {
+         std::string time_string = "0:12:04";
+         auto t = otime::RationalTime(24 * (12 * 60 + 4), 24);
+         auto time_obj = otime::RationalTime::from_time_string(time_string, 24);
+         assertTrue(t.almost_equal(time_obj, 0.001));
```
Comment (jminor) [inline tests/test_opentime.cpp:45]: Could you add a test for negative values, and another to test an invalid input?

**Review by jminor — CHANGES_REQUESTED** [review]

File: src/opentime/rationalTime.cpp [hunk 2]
```diff
@@ -94,6 +94,63 @@ is_dropframe_rate(double rate)
...
+        uintPart = uintPart * 10 + c - '0';
+        ++pCurr;
+    }
+
+    ret = (double) uintPart;
+
+    // check for fractional part
+    if (pCurr == pEnd || *pCurr != '.') {
+        *result = sign * ret;
+        return true;   // no decimal, not a float, stop parsing
```
[the remark this answers is not in the document]
Comment (meshula) [reply src/opentime/rationalTime.cpp:134]: That's a good point.

File: tests/test_opentime.cpp (hunk 1 above)
Comment (meshula) [reply tests/test_opentime.cpp:45]: I should add assertions that `assertTrue` for invalid time returns for all the error conditions that are tested for in the parser...

File: src/opentime/rationalTime.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (meshula) [reply src/opentime/rationalTime.cpp]: fixed

File: tests/test_opentime.cpp (hunk 1 above)
Comment (meshula) [reply tests/test_opentime.cpp:45]: added a test to raise a value error for a negative time code, and also pink elephants.

File: src/opentime/rationalTime.cpp [hunk 3]
```diff
@@ -94,6 +94,69 @@ is_dropframe_rate(double rate)
...
+    }
+
+    double ret = 0.0;
+    double sign = 1.0;
+
+    if (*pCurr == '+')
+        ++pCurr;
+    else if (*pCurr == '-') {
+        if (!allow_negative) {
+            *result = 0.f;
```
Comment (jminor) [inline src/opentime/rationalTime.cpp:112]: Some places in this code use `0.f` and others `0.0` is there a meaningful difference?

File: src/opentime/rationalTime.cpp [hunk 4]
```diff
@@ -94,6 +94,69 @@ is_dropframe_rate(double rate)
...
+    if (pCurr >= pEnd || !pCurr) {
+        *result = 0.0;
+        return false;
+    }
+
+    double ret = 0.0;
+    double sign = 1.0;
+
+    if (*pCurr == '+')
+        ++pCurr;
```
Comment (jminor) [inline src/opentime/rationalTime.cpp:109]: Do we have a code convention/opinion on mixing braces or non-braces?

File: src/opentime/rationalTime.cpp [hunk 5]
```diff
@@ -94,6 +94,69 @@ is_dropframe_rate(double rate)
...
+
+    // if the next character is not a decimal point, the string is malformed.
+    if (*pCurr != '.') {
+        *result = 0.f; // zero consistent with earlier error condition
+        return false;
+    }
+
+    ++pCurr; // skip decimal
+
+    double scaler = 0.1;
```
Comment (jminor) [inline src/opentime/rationalTime.cpp:145]: On first read, I was stuck on whether this is a "scalar" or a scale factor. I see that it is the order of magnitude, or place-value scale. Maybe just a comment would help others avoid the confusion?

File: src/opentime/rationalTime.cpp [hunk 6]
```diff
@@ -205,55 +268,88 @@ RationalTime::from_timecode(
...
+        if (*current == ':') {
+            parse_end = current + 1;
+            char c = *parse_end;
+            if (c != '\0' && c != ':') {
+                if (c< '0' || c > '9') {
+                    set_error(time_string, ErrorStatus::INVALID_TIME_STRING, error_status);
+                    return RationalTime::_invalid_time;
+                }
+                double val = 0.0;
+                parseFloat(parse_end, prev_parse_end + 1, false, &val);
```
Comment (jminor) [inline src/opentime/rationalTime.cpp:318]: The success/fail return value of parseFloat should be checked and turned into a set_error call, right? Same thing for the 2nd use of parseFloat below.

**Review by jminor — CHANGES_REQUESTED** [review]

File: src/opentime/rationalTime.cpp [hunk 7]
```diff
@@ -94,6 +94,69 @@ is_dropframe_rate(double rate)
...
+    }
+
+    // get integer part
+    uint32_t uintPart = 0;
+    while (pCurr < pEnd) {
+        char c = *pCurr;
+        if (c < '0' || c > '9')
+            break;
+        uintPart = uintPart * 10 + c - '0';
+        ++pCurr;
```
Comment (markreidvfx) [inline src/opentime/rationalTime.cpp:126]: shouldn't it also be a error here if `uintPart` this overflows?

File: src/opentime/rationalTime.cpp (hunk 3 above)
Comment (meshula) [reply src/opentime/rationalTime.cpp:112]: 0.f is a floating point constant, 0.0 is a double. It should be 0.0 here, although it won't affect codegen.

File: src/opentime/rationalTime.cpp (hunk 4 above)
Comment (meshula) [reply src/opentime/rationalTime.cpp:109]: the rest of the file is as bracey as possible, and new lines before braces. I'll update

File: src/opentime/rationalTime.cpp (hunk 5 above)
Comment (meshula) [reply src/opentime/rationalTime.cpp:145]: I've renamed it "position_scale" for unambiguity.

File: src/opentime/rationalTime.cpp (hunk 6 above)
Comment (meshula) [reply src/opentime/rationalTime.cpp:318]: done

File: src/opentime/rationalTime.cpp (hunk 7 above)
Comment (meshula) [reply src/opentime/rationalTime.cpp:126]: That had not occurred to me. I added an overflow check

File: src/opentime/rationalTime.cpp [hunk 8]
```diff
@@ -205,55 +288,108 @@ RationalTime::from_timecode(
...
+static void
+set_error(std::string const& time_string,
+          ErrorStatus::Outcome code,
+          ErrorStatus* err)
+{
+    if (err) {
+        *err = ErrorStatus(
+            code,
+            string_printf(
+                "Input time string '%s' is an invalid time string",
```
Comment (ssteinbach) [inline src/opentime/rationalTime.cpp:300]: Should either this message be more generic, or this specific message passed in?  Or maybe the function name made more specific?  It seems very focused on invalid time string but has the name `set_error`

**Review by ssteinbach — COMMENTED** [review]
Looks good. One little question about naming. Like to see those c++ tests start to fill in!

File: src/opentime/rationalTime.cpp (hunk 4 above)
Comment (rogernelson) [reply src/opentime/rationalTime.cpp:111]: There's a `.clang-format` file in the root of the repo.  Running clang-format should confirm everything to the same standard.

File: src/opentime/rationalTime.cpp [hunk 9]
```diff
@@ -94,6 +94,103 @@ is_dropframe_rate(double rate)
...
+        {
+            // if there are too many digits, resulting in an overflow, fail
+            *result = 0.0;
+            return false;
+        }
+        uintPart = accumulated;
+        ++pCurr;
+    }
+
+    ret = (double) uintPart;
```
Comment (rogernelson) [inline src/opentime/rationalTime.cpp:151]: You might want to consider a static_cast here instead of a C cast -- static_casts are verified at compile team and C casts at runtime (and can thus fail).

File: src/opentime/rationalTime.cpp (hunk 8 above)
Comment (meshula) [reply src/opentime/rationalTime.cpp]: ok, made set_error generic

File: src/opentime/rationalTime.cpp (hunk 9 above)
Comment (meshula) [reply src/opentime/rationalTime.cpp]: sounds good, done

**Review by ssteinbach — APPROVED** [review]
Lgtm, thanks!

**Review by jminor — APPROVED** [review]
